ivc/src/poseidon_quadri/interpreter.rs

```diff
@@ -0,0 +1,204 @@
+    F: PrimeField,
+    const STATE_SIZE: usize,
+    const NB_FULL_ROUND: usize,
+    PARAMETERS,
+    Env,
+>(
+    env: &mut Env,
+    param: &PARAMETERS,
+) -> [Env::Variable; STATE_SIZE]
+where
+    F: PrimeField,
+    PARAMETERS: PoseidonParams<F, STATE_SIZE, NB_FULL_ROUND>,
+    Env: ColAccessCap<F, PoseidonColumn<STATE_SIZE, NB_FULL_ROUND>>
+        + HybridCopyCap<F, PoseidonColumn<STATE_SIZE, NB_FULL_ROUND>>,
+{
+    // Checking that p - 1 is coprime with 7 as it has to be the case for the sbox
+    {
+        let one = BigUint::from(1u64);
+        let p: BigUint = TryFrom::try_from(<F as PrimeField>::Params::MODULUS).unwrap();
+        let p_minus_one = p - one.clone();
+        let seven = BigUint::from(7u64);
+        assert_eq!(p_minus_one.gcd(&seven), one);
+    }
+
+    let mut final_state: [Env::Variable; STATE_SIZE] =
+        std::array::from_fn(|_| Env::constant(F::zero()));
+
+    for i in 0..NB_FULL_ROUND {
+        let state: [PoseidonColumn<STATE_SIZE, NB_FULL_ROUND>; STATE_SIZE] = {
+            if i == 0 {
+                std::array::from_fn(PoseidonColumn::Input)
+            } else {
+                let prev_round = i - 1;
+                // Previous outputs are in index 4, 9, and 14 if we have 3 elements
+                std::array::from_fn(|j| PoseidonColumn::Round(prev_round, j * 5 + 4))
+            }
+        };
+        let round_res = compute_one_round::<F, STATE_SIZE, NB_FULL_ROUND, PARAMETERS, Env>(
+            env, param, i, &state,
+        );
+
+        if i == NB_FULL_ROUND - 1 {
+            final_state = round_res
+        }
+    }
+
+    final_state
+}
+
+/// Compute one round the Poseidon permutation
+fn compute_one_round<
+    F: PrimeField,
+    const STATE_SIZE: usize,
+    const NB_FULL_ROUND: usize,
+    PARAMETERS,
+    Env,
+>(
+    env: &mut Env,
+    param: &PARAMETERS,
+    round: usize,
+    elements: &[PoseidonColumn<STATE_SIZE, NB_FULL_ROUND>; STATE_SIZE],
+) -> [Env::Variable; STATE_SIZE]
+where
+    F: PrimeField,
+    PARAMETERS: PoseidonParams<F, STATE_SIZE, NB_FULL_ROUND>,
+    Env: ColAccessCap<F, PoseidonColumn<STATE_SIZE, NB_FULL_ROUND>>
+        + HybridCopyCap<F, PoseidonColumn<STATE_SIZE, NB_FULL_ROUND>>,
+{
+    // We start at round 0
+    // This implementation mimicks the version described in
+    // poseidon_block_cipher in the mina_poseidon crate.
+    assert!(
+        round < NB_FULL_ROUND,
+        "The round index {:} is higher than the number of full rounds encoded in the type",
+        round
+    );
+    // Applying sbox
+    // For a state transition from (x, y, z) to (x', y', z'), we use the
+    // following columns shape:
+    // x^2, x^4, x^6, x^7, x', y^2, y^4, y^6, y^7, y', z^2, z^4, z^6, z^7, z')
+    //  0    1    2    3   4   5    6    7    8    9    10   11   12   13  14
+    let state: Vec<Env::Variable> = elements
+        .iter()
+        .enumerate()
+        .map(|(i, x_col)| {
+            let x = env.read_column(*x_col);
```

Comment:
tiny nit: I'd `s/x/var/` in the whole block since it's not just `x` but `x,y,z`.

Reply:
Done